warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeHashVisitor.java

```diff
@@ -117,38 +118,39 @@ public Object visit(ASTReferenceExpression node, Object data) {
     }
     
     public Object visit(ASTOrNode node, Object data) {
-        
-        int numChildren = node.jjtGetNumChildren();
-        
-        TreeHashNode inte = (TreeHashNode) (data);
-        int lastsize = inte.length();
-        
-        for (int i = 0; i < numChildren; i++) {
-            if (inte.length() != lastsize) {
-                inte.append(ORNODE);
-            }
-            lastsize = inte.length();
-            node.jjtGetChild(i).jjtAccept(this, inte);
-        }
-        
-        return inte;
+        return visitAndOrChildren(node, data, ORNODE);
     }
     
     public Object visit(ASTAndNode node, Object data) {
-        
+        return visitAndOrChildren(node, data, ANDNODE);
+    }
+    
+    private Object visitAndOrChildren(JexlNode node, Object data, String joinTerm) {
         int numChildren = node.jjtGetNumChildren();
         
         TreeHashNode inte = (TreeHashNode) (data);
         int lastsize = inte.length();
         
-        for (int i = 0; i < numChildren; i++) {
+        // Build keys for child node map
+        List<String> keys = new ArrayList<>();
+        TreeMap<String,JexlNode> nodeMap = new TreeMap<>();
+        for (int ii = 0; ii < numChildren; ii++) {
+            JexlNode child = node.jjtGetChild(ii);
+            String key = TreeHashVisitor.getNodeHash(child).toString();
+            nodeMap.put(key, child);
```

Comment:
Would the TreeNodeHash object serve the purpose here?  When the toString()?

Reply:
The TreeNodeHash.toString() takes the integer hash and appends the nodes. This protects from two different sub trees that hash to the same value.

Reply:
We should not be depending on toString() for functionality as such.  I would much rather use the TreeNodeHash object itself.  Using toString can be dangerous as it is not guarenteed to be a unique representation from a Java point of view.

Reply:
Fixed this by having the TreeHashNode object implement comparable

Reply:
now were talking.  thanks.